src/main/groovy/com/avast/gradle/dockercompose/ComposeExecutor.groovy

```diff
@@ -0,0 +1,125 @@
+package com.avast.gradle.dockercompose
+
+import org.gradle.api.Project
+import org.gradle.api.logging.Logger
+import org.gradle.process.ExecSpec
+import org.gradle.util.VersionNumber
+import org.yaml.snakeyaml.Yaml
+
+import java.util.concurrent.Executors
+
+class ComposeExecutor {
+    private final ComposeExtension extension
+    private final Project project
+    private final Logger logger
+
+    ComposeExecutor(ComposeExtension extension) {
+        this.extension = extension
+        this.project = extension.project
+        this.logger = extension.project.logger
+    }
+
+    private void executeWithCustomOutput(OutputStream os, String... args) {
+        def ex = this.extension
+        project.exec { ExecSpec e ->
+            if (extension.dockerComposeWorkingDirectory) {
+                e.setWorkingDir(extension.dockerComposeWorkingDirectory)
+            }
+            e.environment = ex.environment
+            def finalArgs = [ex.executable]
+            finalArgs.addAll(ex.useComposeFiles.collectMany { ['-f', it] })
+            if (ex.projectName) {
+                finalArgs.addAll(['-p', ex.projectName])
+            }
+            finalArgs.addAll(args)
+            e.commandLine finalArgs
+            e.standardOutput = os
+        }
+    }
+
+    String execute(String... args) {
+        new ByteArrayOutputStream().withStream { os ->
+            executeWithCustomOutput(os, args)
+            os.toString().trim()
+        }
+    }
+
+    VersionNumber getVersion() {
+        VersionNumber.parse(execute('--version').findAll(/(\d+\.){2}(\d+)/).head())
```

Comment:
it looks to me that version is accessed many times when executing docker compose - maybe it makes sense to compute it and store it? (it does not change during the run of the plugin)

Reply:
You're right, it's not necessary now. But I'm going to implement #36 soon and then it will be possible to use different docker-compose.

I will see how the design will be and consider caching of this value.